pkg/gce-cloud-provider/compute/gce-compute.go

```diff
@@ -456,3 +474,116 @@ func (cloud *CloudProvider) GetInstanceOrError(ctx context.Context, instanceZone
+	glog.V(4).Infof("Getting snapshot %v", snapshotName)
+	snapshot, err := svc.Snapshots.Get(project, snapshotName).Context(ctx).Do()
+	if err != nil {
+		return nil, err
+	}
+	glog.V(4).Infof("Got snapshot %v", snapshotName)
+	return snapshot, nil
+}
+
+func (cloud *CloudProvider) DeleteSnapshot(ctx context.Context, snapshotName string) error {
+	op, err := cloud.service.Snapshots.Delete(cloud.project, snapshotName).Context(ctx).Do()
+	if err != nil {
+		if IsGCEError(err, "notFound") {
+			// Already deleted
+			return nil
+		}
+		return err
+	}
+	err = cloud.waitForGlobalOp(ctx, op)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cloud *CloudProvider) CreateSnapshot(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	switch volKey.Type() {
+	case meta.Zonal:
+		snapshot, err := cloud.createSnapshotOfZonalDisk(ctx, volKey, snapshotName)
+		return snapshot, err
+	case meta.Regional:
+		snapshot, err := cloud.createSnapshotOfRegionalDisk(ctx, volKey, snapshotName)
+		return snapshot, err
+	default:
+		return nil, fmt.Errorf("could not create snapshot, key was neither zonal nor regional, instead got: %v", volKey.String())
+	}
+}
+
+func (cloud *CloudProvider) createSnapshotOfZonalDisk(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	snapshotToCreate := &compute.Snapshot{
+		Name:        snapshotName,
+	}
+
+	_, err := cloud.service.Disks.CreateSnapshot(cloud.project, volKey.Zone, volKey.Name, snapshotToCreate).Context(ctx).Do()
+
+	if err != nil {
+		if IsGCEError(err, "alreadyExists") {
+			snapshot, err := cloud.GetSnapshot(ctx, snapshotName)
+			if err != nil {
+				return nil, err
+			}
+			glog.Warningf("GCE PD snapshot %s already exists, reusing", snapshotName)
+			return snapshot, nil
+		}
+		return nil, fmt.Errorf("unkown disk create snapshot error: %v", err)
+	}
+
+	return cloud.waitForSnapshotCreation(ctx, snapshotName)
+}
+
+func (cloud *CloudProvider) createSnapshotOfRegionalDisk(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	snapshotToCreate := &computebeta.Snapshot{
+		Name:        snapshotName,
+	}
+
+	_, err := cloud.betaService.RegionDisks.CreateSnapshot(cloud.project, volKey.Region, volKey.Name, snapshotToCreate).Context(ctx).Do()
+	if err != nil {
+		if IsGCEError(err, "alreadyExists") {
+			snapshot, err := cloud.GetSnapshot(ctx, snapshotName)
+			if err != nil {
+				return nil, err
+			}
+			glog.Warningf("GCE PD snapshot %s already exists, reusing", snapshotName)
+			return snapshot, nil
+		}
+		return nil, fmt.Errorf("unkown disk create snapshot error: %v", err)
+	}
+
+	return cloud.waitForSnapshotCreation(ctx, snapshotName)
+
+}
+
+func (cloud *CloudProvider) waitForSnapshotCreation(ctx context.Context, snapshotName string) (*compute.Snapshot, error) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	timer := time.NewTimer(waitForSnapshotCreationTimeOut)
```

Comment:
curious as to why we choose 2 minutes here and why we use a separate timeout instead of the parent context

Reply:
Because normally it only takes seconds to create snapshot, two minutes should be enough. which parent context you mean?